flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/ExplodeFunctionUtil.scala

```diff
@@ -18,74 +18,143 @@
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Object, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
 class FloatExplodeTableFunc extends TableFunction[Float] {
   def eval(arr: Array[Float]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Float, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
 class ShortExplodeTableFunc extends TableFunction[Short] {
   def eval(arr: Array[Short]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Short, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 class IntExplodeTableFunc extends TableFunction[Int] {
   def eval(arr: Array[Int]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Int, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
 class LongExplodeTableFunc extends TableFunction[Long] {
   def eval(arr: Array[Long]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Long, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
 class DoubleExplodeTableFunc extends TableFunction[Double] {
   def eval(arr: Array[Double]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Double, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
 class ByteExplodeTableFunc extends TableFunction[Byte] {
   def eval(arr: Array[Byte]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Byte, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
 class BooleanExplodeTableFunc extends TableFunction[Boolean] {
   def eval(arr: Array[Boolean]): Unit = {
-    arr.foreach(collect)
+    CommonCollect.collectArray(arr, collect)
+  }
+
+  def eval(map: util.Map[Boolean, Integer]): Unit = {
+    CommonCollect.collect(map, collect)
   }
 }
 
-object ExplodeFunctionUtil {
-  def explodeTableFuncFromType(ti: TypeInformation[_]):TableFunction[_] = {
-    ti match {
-      case pat: PrimitiveArrayTypeInfo[_] => {
-        pat.getComponentType match {
-          case BasicTypeInfo.INT_TYPE_INFO => new IntExplodeTableFunc
-          case BasicTypeInfo.LONG_TYPE_INFO => new LongExplodeTableFunc
-          case BasicTypeInfo.SHORT_TYPE_INFO => new ShortExplodeTableFunc
-          case BasicTypeInfo.FLOAT_TYPE_INFO => new FloatExplodeTableFunc
-          case BasicTypeInfo.DOUBLE_TYPE_INFO => new DoubleExplodeTableFunc
-          case BasicTypeInfo.BYTE_TYPE_INFO => new ByteExplodeTableFunc
-          case BasicTypeInfo.BOOLEAN_TYPE_INFO => new BooleanExplodeTableFunc
+object CommonCollect {
+
+  def collectArray[T](array: Array[T], collectFunc: (T) => Unit): Unit = {
```

Comment:
`collectFunc: (T) => Unit` is another Scala magic where we don't know how this is translated into byte code.